### PR TITLE
Node-ip is not used when cloud provider is used

### LINKED ISSUE #23568: cloud.Instance.NodeAddress() --node-ip to generate list of node addresses
https://github.com/kubernetes/kubernetes/pull/18541 adds a node-ip flag to kubelet.  However, it's only used for non-cloudprovider case.  The same can also happen in a cloud environment (e.g. if you want to form a private network in the cloud).   

Passing --node-ip to NodeAddress() can give hint to cloudprovider plugin.

### PR DESCRIPTION
**Pull request by kokhang**
Currently --node-ip in kubelet is not being used when kubelet is configured with a cloud provider. With this fix, kubelet will get a list of IPs from the provider and parse it to return the one that matches node-ip.

This fixes #23568

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33123)

### CODE REVIEW

**Comment by dagnello** [comment]
hello @yujuhong, can we get tests run for this?  thanks

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by kokhang** [comment]
@yujuhong GCE e2e test failed but it doesnt seem to be because of the patch. I ran `make test WHAT="pkg/kubelet"` locally and that passed. Do you know why it failed? If not, could you run it again? Thanks

**Comment by yujuhong** [comment]
@k8s-bot gce e2e test this

**Comment by ddysher** [comment]
@kokhang I think we are facing a slightly different case.  In my case, I want to pass nodeIP to cloud provider plugin and let it decide whether to include nodeIP to list of nodeAddresses, and what type it is.

File: pkg/kubelet/kubelet_node_status.go [hunk 1]
```diff
@@ -372,6 +372,19 @@ func (kl *Kubelet) setNodeAddress(node *api.Node) error {
 		if err != nil {
 			return fmt.Errorf("failed to get node address from cloud provider: %v", err)
 		}
+
+		if kl.nodeIP != nil {
+			for _, nodeAddress := range nodeAddresses {
+				if nodeAddress.Address == kl.nodeIP.String() {
```
Comment (ddysher) [inline pkg/kubelet/kubelet_node_status.go]: what if no nodeIP is passed from NodeAddress but I still want to include it?

File: pkg/kubelet/kubelet_node_status.go [hunk 2]
```diff
@@ -372,6 +372,19 @@ func (kl *Kubelet) setNodeAddress(node *api.Node) error {
...
+					node.Status.Addresses = []api.NodeAddress{
+						{Type: nodeAddress.Type, Address: nodeAddress.Address},
```
Comment (ddysher) [inline pkg/kubelet/kubelet_node_status.go]: You have to give it a type?

**Comment by kokhang** [comment]
@ddysher I was thinking about that but giving the node-ip to the cloud provider would required a lot of changes. Since we would need to implement this in all cloud provider.

File: pkg/kubelet/kubelet_node_status.go (hunk 1 above)
Comment (kokhang) [inline pkg/kubelet/kubelet_node_status.go]: Would you just add this node-ip to the list of addresses even if its not part of the list?
The reason i am searching through the list of returned IPs is to make sure that the node-ip can be routable to the node.

File: pkg/kubelet/kubelet_node_status.go (hunk 2 above)
Comment (kokhang) [inline pkg/kubelet/kubelet_node_status.go]: I am providing a type in `Type: nodeAddress.Type,`

**Comment by yujuhong** [comment]
@ddysher @kokhang what's the use case of this?

**Comment by kokhang** [comment]
@yujuhong This is for administrator to choose an IP for that node in the case there the nodes has multiple IPs. Look at https://github.com/kubernetes/kubernetes/pull/18541. 

However the fix for that issue was only applied for kubelet that were not configured with a cloud provider. This PR makes sure that node-ip also applies when kubelet is configured wit ha cloud pro vider.

**Comment by kokhang** [comment]
@yujuhong Any other question about this PR? Thanks

**Comment by yujuhong** [comment]
@kokhang I see the point, but I'm a bit concerned about the inconsistent behavior when using node IP.
Without the cloud provider, kubelet [validate the node IP and fail if none exist](https://github.com/kubernetes/kubernetes/blob/8ef3c8b389919f4bab720eb683667221fe3cb44d/pkg/kubelet/kubelet.go#L466). In this PR, kubelet will simply not come up. But with your change, kubelet will just not register with the apiserver.  Could you move the `validateNodeIP` function to `setNodeAddress()`and clean up the logic? Thanks

**Comment by kokhang** [comment]
@k8s-bot kubemark e2e test this

**Comment by kokhang** [comment]
@k8s-bot gci gke e2e test this

**Comment by kokhang** [comment]
@k8s-bot gci gce e2e test this

**Comment by kokhang** [comment]
@yujuhong it seems some of the tests failed but i dont think it is related to my PR. I tried to tell @k8s-bot to test again but i dont think i have the privilege to do that. Can you re-run the test again? Thanks

**Comment by yujuhong** [comment]
LGTM.

**Comment by yujuhong** [comment]
@k8s-bot gci gce e2e test this

**Pull request merged**